[AIRFLOW-5162] GCS Hook Upload Method Improvement

steven-freed opened the pull request:
**Jira**

- [x] My PR addresses the following Airflow Jira issues and references them in the PR title. For example, "[AIRFLOW-XXX] My Airflow PR"
- https://issues.apache.org/jira/browse/AIRFLOW-5162

**Description**
- [x] Here are some details about my PR, including screenshots of any UI changes:
- Allows developers to upload objects to a GCS bucket by either a local file's file name or by a file's data (string or bytes). This may help prevent issues related to crash consistency as well as help developers have a larger variety of uploading options to GCS.

**Tests**
- [x] My PR adds the following unit tests OR does not need testing for this extremely good reason:
- The changes made to the 'upload' method still allow for old functionality, while supporting new functionality for Google GCS API methods. 

**Commits**
- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "How to write a good git commit message":

Subject is separated from body by a blank line
Subject is limited to 50 characters (not including Jira issue reference)
Subject does not end with a period
Subject uses the imperative mood ("add", not "adding")
Body wraps at 72 characters
Body explains "what" and "why", not "how"

**Documentation**
- In case of new functionality, my PR adds documentation that describes how to use it.
All the public functions and the classes in the PR contain docstrings that explain what it does
If you implement backwards incompatible changes, please leave a note in the Updating.md so we can assign it to a appropriate release

**Code Quality**
- [x] Passes flake8

ryanyuan commented on airflow/contrib/hooks/gcs_hook.py:
```diff
@@ -175,7 +175,30 @@ def download(self, bucket_name, object_name, filename=None):
 
         return blob.download_as_string()
 
-    def upload(self, bucket_name, object_name, filename,
```
Is this going to break the operators that are calling gcs_hook.upload()?

steven-freed replied:
Yes it would because the new version would replace 'upload' with 'upload_from_filename' and 'upload_from_string'. That is a very good point thank you. I have modified the upload method to be more generic for both types of uploads.

Commit:
[AIRFLOW-5162] add to 'upload' method for gcs file data

Allows for local file uploads (parameter 'filename') and file content (parameter 'data') uploads to gcs by using Google's gcs api method 'upload_from_string' which allows for uploading files content as a string or bytes.

mik-laj commented on airflow/contrib/hooks/gcs_hook.py:
```diff
@@ -175,40 +175,55 @@ def download(self, bucket_name, object_name, filename=None):
         :param mime_type: The MIME type to set when uploading the file.
         :type mime_type: str
-        :param gzip: Option to compress file for upload
+        :param gzip: Option to compress local file for upload
```
Why didn't you add compression support for data?
```
            if gzip:
                out = StringIO.StringIO()
                with gz.GzipFile(fileobj=out, mode="w") as f:
                    f.write(data)
                data = out.getvalue()
```

steven-freed replied:
I actually had a use case at work where compression couldn’t be used when uploading files to BigQuery from a Storage Bucket. I didn’t think to add it but that is a great idea. I will add it asap. Thank you

Commit:
[AIRFLOW-5162] add ValueError exceptions and upload tests

ValueError exceptions are raised when the filename and data params are
provided or neither are provided. This avoids errors of local files and
file data. Tests assert file and data upload variants as well as assert
exceptions.

Commit:
Merge branch 'master' of github.com:apache/airflow into gcs-hook-update

Commit:
[AIRFLOW-5162] fixed unit tests for gcs uploads

added seperate unit tests for uploading string content and byte content
using the upload_from_string gcp method. Prevents assertion being called
twice when tests are seperated as well as organizes testing.

potiuk commented on tests/contrib/hooks/test_gcs_hook.py:
```diff
@@ -602,21 +604,103 @@ def test_upload(self, mock_service):
+                                        test_object,
+                                        data=self.testdata_bytes,
+                                        gzip=True)
+
```
I think we need to run asserr_called_once_with here.

potiuk commented on tests/contrib/hooks/test_gcs_hook.py:
```diff
@@ -602,21 +604,103 @@ def test_upload(self, mock_service):
+            .blob.return_value.upload_from_string
+        upload_method.return_value = None
+
+        response = self.gcs_hook.upload(test_bucket,  # pylint:disable=assignment-from-no-return
```
I think it's not really needed to add response and assertIsNone() (and add #pylint exceptions). It does not add any value. What you should do instead is to add type annotation for the upload_method() and add -> None there and it will be much better way to make sure it behaves as it should (via static checks rather than tests). Adding type annotations is a good thing actually

potiuk commented on airflow/contrib/hooks/gcs_hook.py:
```diff
@@ -185,40 +186,64 @@ def download(self, bucket_name, object_name, filename=None):
 
-    def upload(self, bucket_name, object_name, filename,
-               mime_type='application/octet-stream', gzip=False):
+    def upload(self, bucket_name, object_name, filename=None,
```
Can you please add type annotations here? see the comment below - that would be much better than assertNone and also it's great for users of the hook.

potiuk commented on tests/contrib/hooks/test_gcs_hook.py:
```diff
@@ -602,21 +604,103 @@ def test_upload(self, mock_service):
+            .blob.return_value.upload_from_string
+        upload_method.return_value = None
+
+        self.assertRaises(ValueError,
```
It's not really nice way of using asserRaises. It is difficult to read. I think it should look similar to:

```
with self.assertRaises(ValueError) as cm:
    self.gcs_hook.upload(test_bucket,test_object)

the_exception = cm.exception
self.assertEqual("... Message here...", the_exception.message)
```

potiuk reviewed (changes requested):
There are still few comments - I think the test code might be much more readable (I proposed a few changes) and some assertions can be skipped in favour of Type Annotation.

Commit:
[AIRFLOW-5162] added type annotations and tests modifications

Added type annotations for static tests and ease of use for users. Tests
modifications get rid of assertions for None responses and add tests for
calling upload functions once.

Commit:
Merge branch 'master' into gcs-hook-update

Commit:
[AIRFLOW-5162] add pylint and flake8 fixes for type annotations

type annotation syntax requires a space between '=' sign for default
assignment.

Commit:
Merge branch 'master' of github.com:apache/airflow into gcs-hook-update

Commit:
[AIRFLOW-5162] added called once and exception tests

added called once tests for upload methods for gzip data. This was added
to assert that the upload_from_string method was called correctly.
Assertions for exceptions confirming the correct exception messages.

mik-laj commented on airflow/gcp/hooks/gcs.py:
```diff
@@ -185,45 +180,72 @@ def download(self, bucket_name, object_name, filename=None):
         """
-        Uploads a local file to Google Cloud Storage.
-
+        Uploads a local file or file data as string or bytes to Google Cloud Storage.
```
New line is required after method description. Can you add it?

mik-laj commented on airflow/gcp/hooks/gcs.py:
```diff
@@ -185,45 +180,72 @@ def download(self, bucket_name, object_name, filename=None):
 
     def exists(self, bucket_name, object_name):
         """
         Checks for the existence of a file in Google Cloud Storage.
```
This change prevents the documentation from being built properly. Can you delete it?

mik-laj commented on airflow/gcp/hooks/gcs.py:
```diff
@@ -238,7 +260,6 @@ def exists(self, bucket_name, object_name):
     def is_updated_after(self, bucket_name, object_name, ts):
         """
         Checks if an blob_name is updated in Google Cloud Storage.
-
```
This change prevents the documentation from being built properly. Can you delete it?

mik-laj commented on airflow/gcp/hooks/gcs.py:
```diff
@@ -273,7 +294,6 @@ def is_updated_after(self, bucket_name, object_name, ts):
     def delete(self, bucket_name, object_name):
         """
         Deletes an object from the bucket.
-
```
This change prevents the documentation from being built properly. Can you delete it?

mik-laj commented on airflow/gcp/hooks/gcs.py:
```diff
@@ -289,7 +309,6 @@ def delete(self, bucket_name, object_name):
     def list(self, bucket_name, versions=None, max_results=None, prefix=None, delimiter=None):
         """
         List all objects from the bucket with the give string prefix in name
-
```
This change prevents the documentation from being built properly. Can you delete it?

mik-laj commented on airflow/gcp/hooks/gcs.py:
```diff
@@ -336,13 +355,11 @@ def list(self, bucket_name, versions=None, max_results=None, prefix=None, delimi
     def get_size(self, bucket_name, object_name):
         """
         Gets the size of a file in Google Cloud Storage.
-
```
This change prevents the documentation from being built properly. Can you delete it?

mik-laj commented on airflow/gcp/hooks/gcs.py:
```diff
@@ -357,7 +374,6 @@ def get_size(self, bucket_name, object_name):
     def get_crc32c(self, bucket_name, object_name):
         """
         Gets the CRC32c checksum of an object in Google Cloud Storage.
-
```
This change prevents the documentation from being built properly. Can you delete it?

mik-laj commented on airflow/gcp/hooks/gcs.py:
```diff
@@ -405,11 +420,9 @@ def create_bucket(self,
         .. seealso::
             For more information, see Bucket Naming Guidelines:
             https://cloud.google.com/storage/docs/bucketnaming.html#requirements
-
```
This change prevents the documentation from being built properly. Can you delete it?

mik-laj commented on airflow/gcp/hooks/gcs.py:
```diff
@@ -405,11 +420,9 @@ def create_bucket(self,
         """
         Creates a new bucket. Google Cloud Storage uses a flat namespace, so
         you can't create a bucket with a name that is already in use.
-
```
This change prevents the documentation from being built properly. Can you delete it?

mik-laj commented on airflow/gcp/hooks/gcs.py:
```diff
@@ -418,23 +431,19 @@ def create_bucket(self,
         :type resource: dict
         :param storage_class: This defines how objects in the bucket are stored
             and determines the SLA and the cost of storage. Values include
-
```
I would also ask all other unrelated changes to the documentation to be withdrawn.

Commit:
Merge branch 'master' of https://github.com/apache/airflow into gcs-hook-update

Commit:
[AIRFLOW-5162] added fix to doc build

Fixed indentation in docstrings so that the documentation can build
properly. Added new lines after descriptions of the methods
functionalities.
